Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -145,8 +164,47 @@ public struct HTTPResponseHead: Equatable {
     }
 }
 
-fileprivate typealias HTTPHeadersStorage = [String: [(String, String)]] // [lowerCasedName: [(originalCaseName, value)]
+/// The Index for a header name or value that points into the underlying `ByteBuffer`.
+struct HTTPHeaderIndex {
+    let start: Int
+    let length: Int
+}
+
+/// Struct which holds name, value pairs.
+struct HTTPHeader {
+    let name: HTTPHeaderIndex
+    let value: HTTPHeaderIndex
+}
+
+private extension ByteBuffer {
+    func equalCaseInsensitiveASCII(view: String.UTF8View, at index: HTTPHeaderIndex) -> Bool {
+        guard view.count == index.length else {
+            return false
+        }
+        return withVeryUnsafeBytes { buffer in
+            // This should never happens as we control when this is called. Adding and assert to ensure this.
+            assert(index.start <= self.capacity - index.length)
```

Comment:
I think I pretty strongly recommend upgrading this to a `precondition`, as the consequences of this failing is memory unsafety.

Reply:
@Lukasa usually we use asserts where something will happen because of an bug in NIO and `precondition` when the user can provide the input. Thats why I used `assert` but if you are strong on this I can change it (it's just not consistent with what we do in other places IMHO). @weissi WDYT ?